clientlibrary/worker/worker.go

```diff
@@ -284,10 +310,23 @@ func (w *Worker) eventLoop() {
 				}
 
 				// The shard is closed and we have processed all records
-				if shard.Checkpoint == chk.SHARD_END {
+				if shard.GetCheckpoint() == chk.SHARD_END {
 					continue
 				}
 
+				var stealShard bool
```

Comment:
Why this block cannot be moved to reblance()?

Reply:
Part of this is an optimization of the inner for loop in the event loop, and part of it is a state machine to force linear convergence via `w.shardStealInProgress`. Because of how Claim Shard and GetLease work together, I feel like it makes sense for it stay where it is. We'd can rethink how we could move it into rebalance. But it will be tricky. Let us know what you want to do.